src/main/java/com/dabsquared/gitlabjenkins/gitlab/api/impl/AutodetectGitLabClientBuilder.java

```diff
@@ -18,11 +18,25 @@ public AutodetectGitLabClientBuilder() {
 
     @Override
     @NonNull
-    public GitLabClient buildClient(
+    public GitLabApi buildClient(
```

Comment:
The original version of this code was more elegant I think. It retrieved the sorted list of all builders (autodetecting, V3, and V4), removed itself (the autodetecting builder) from the list, and then delegated to the existing implementation in the V3 and V4 builders. Your version duplicates the implementation in the autodetecting builder. It would be more elegant to reuse the existing code in the V3 and V4 builders. This can be done in this method with something like this:

```java
    @Override
    @NonNull
    public GitLabApi buildClient(String url, String token, boolean ignoreCertificateErrors, int connectionTimeout, int readTimeout) {
        Collection<GitLabClientBuilder> candidates = new ArrayList<>(getAllGitLabClientBuilders());
        candidates.remove(this);
        return autodetectOrDie();
    }

    @NonNull
    private GitLabApi autodetectOrDie(Collection<GitLabClientBuilder> candidates) {
        GitLabApi client = autodetect(candidates);
        if (client != null) {
            return client;
        }

        throw new NoSuchElementException("no client-builder found that supports server at " + url);
    }

    @CheckForNull
    private GitLabApi autodetect(Collection<GitLabClientBuilder> candidates) {
        for (GitLabClientBuilder candidate : candidates) {
            GitLabApi client =
                    candidate.buildClient(url, token, ignoreCertificateErrors, connectionTimeout, readTimeout);
            try {
                client.getCurrentUser();
                return client;
            } catch (<whatever type getCurrentUser would throw if the user was not found> ignored) {
                // api-endpoint not found (== api-level not supported by this client)
            }
        }

        return null;
    }
```